File: CallbackSnippet.java
package me.oogh.demo;

import java.util.ArrayList;
import java.util.List;

/**
 * 工厂类
 */
class Factory {
    public static List<String> generateData() {
        List<String> dataSet = new ArrayList<>();
        dataSet.add("data1");
        dataSet.add("data2");
        dataSet.add("data3");
        dataSet.add("data4");
        return dataSet;

    }
}

/**
 * 测试类
 */
class Repository {
    public void loadData(OnDataLoadedCallback callback) {
        callback.onDataLoaded(me.oogh.demo.Factory.generateData());
    }

    // 定义 Callback
    interface OnDataLoadedCallback {
        void onDataLoaded(List<String> dataSet);
    }
}

/**
 * 销售商类
 */
class Seller {
    private me.oogh.demo.Repository repository = new me.oogh.demo.Repository();

    public void sell() {
        repository.loadData(new me.oogh.demo.Repository.OnDataLoadedCallback() {
            @Override
            public void onDataLoaded(List<String> dataSet) {
                for (String item : dataSet) {
                    System.out.println(item);
                }
            }
        });
    }
}

/**
 * 运行，测试
 */
public class CallbackSnippet {
    public static void main(String[] args) {
        me.oogh.demo.Seller seller = new me.oogh.demo.Seller();
        seller.sell();

    }
}

File: CallbackSnippet.kt
package me.oogh.demo

/**
 * 工厂类
 * 生成数据
 */
class Factory {
    // companion object: 伴随对象，类似于Java中的Static，因此其中的方法可以使用类名直接调用
    companion object {
        fun generateData(): List<String> {
            return listOf("data1", "data2", "data3", "data4")
        }
    }
}

/**
 * 仓库类
 *  从工厂中拿到数据后，返回给商人
 */
class Repository {
    fun loadData(callback: (dataSet: List<String>) -> Unit) {
        callback.invoke(Factory.generateData())
    }
}

/**
 * 销售商
 *  从仓库中拿到数据后，进行销售
 */
class Seller {
    private val repository = Repository()
    fun sell() {
        repository.loadData { dataSet -> dataSet.forEach { println(it) } }
    }
}

/**
 * 运行，测试
 */
fun main(args: Array<String>) {
    val seller = Seller()
    seller.sell()
}